src/boot/efifirmware.h

```diff
@@ -0,0 +1,29 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+#pragma once
+
+#include "efi.h"
+
+#define FWHEADERMAGIC (UINT32_C(0xfeeddead))
+
+/*  The structure of the efifw UKI blob is the following:
+ *  ---------------------------------------------------------
+ *  EfiFw header|fwid|payload| reserved for future attributes
+ *  ---------------------------------------------------------
+ *  The base header defines the length of full header, fwid and payload.
+ *  The fwid is a NULL terminated string.
+ *  The payload contains the actual efi firmware.
+ */
+typedef struct EfiFwHeader {
+        struct {
+                uint32_t magic;
+                uint32_t header_len; /* includes base + additional attributes */
+                uint32_t fwid_len;
+                uint32_t payload_len;
+        } base; /* base attributes */
```

Comment:
i'd drop the `base` substruct tbh, just make the fields part of the main EfiFwHeader structure

Reply:
I prefer to leave this as is for now. base attributes in a separate struct looks cleaner?

Reply:
why have this extra nested struct at all? we have no other attributes for now, and we have nothing planned any time soon. once we add more attributes we can certainly move this down, but right now, i really see no reason to.

in particular, as when the day comes we cannot just extend EfiFwHeader as it is with additional fields, because then we couldn't use the struct for old files anymore (that do not carry the extra fields), because in C it's by specification require that structure pointers dereference to entirely valid memory from the first to the last byte of the struct, and hence older short headers in existing files could not be covered by the new extended structure definition.

Reply:
> once we add more attributes we can certainly move this down
but if we start rearranging the struct, we will lost compatiblity with existing users who do not know about the new header. That is why I wanted to make sure base struct remains as is in all versions of the header and additional members are only added below. That way in all versions, base is guranteed to be present. The presence of new attributes can be checked by checking the length of the header.